### PR TITLE
Show transforms as a vertical list

### LINKED ISSUE #22073: Show transforms as a vertical list rather than a grid
Let's try showing block transformations as a list (the same way they render in the `/` inserter) instead of the current grid pattern as it'd help with familiarity with other editors:

image

And this instead:

image

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description

Resolves #22073

In this PR:

- Transforms are shown as a vertical list, so as to help with familiarity with other editors.
- Transforms and Block Styles are combined in a drop down menu and arrow navigation is supported.
- Styles have been moved below transforms.
- Previews have been implemented with a Popover and changed style.
- Added support for `role` attribute in `BlockStyles` component.
- Enhanced Styles Preview at smaller viewport ( 782px ), when admin bar goes big

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by youknowriad** [comment]
Cool stuff here Nick.

I'm testing on safari and I  noticed two small things:

- There's some kind of border under the first item.

Capture d’écran 2020-06-09 à 3 47 49 PM

- Feels like the padding around the whole dropdown is bigger than the other popovers?

**Comment by ntsekouras** [comment]
> There's some kind of border under the first item.

I will fix that. It's the focused style I see.

> Feels like the padding around the whole dropdown is bigger than the other popovers?

Yes, That was the padding from before (grid). I will change it to match the other vertical lists padding.

Thanks @youknowriad!

**Review by youknowriad — COMMENTED** [review]
This is looking good to me overall but there's something I was wondering.

We're not using a DropdownMenu  here and  instead we use just Dropdown, that's fine cause sometimes you want  extra things in the content (like the previews),  but we lose some DropdownMenu features like arrow navigation. it makes me wonder if we should  wrap the  MenuGroups inside NavigableToolbar to enable the arrow navigation.

**Comment by jasmussen** [comment]
This is nice! Thanks for working on this. Here's what I see:

after

At the end, instead of using tab to navigate the menu items, I use arrowkeys, which breaks. That feels a bit frustrating, because arrowkeys work, for example, in the ellipsis menu. This would be nice to address.

Designwise, I think we should try to move closer to the mockup shown in https://github.com/WordPress/gutenberg/issues/18667#issuecomment-559259876:



It doesn't have to be part of this PR — but it's worth sharing the source reason for making into a list. One small step towards that could be to change this:

```
	.block-editor-block-switcher__label {
		margin-bottom: $grid-unit-10;
		color: $medium-gray-text;
	}
```

to this

```
	.block-editor-block-switcher__label {
		margin-bottom: $grid-unit-15;
	}
```

That would get us very close even just with the list.

**Comment by ntsekouras** [comment]
> We're not using a DropdownMenu here and instead we use just Dropdown, that's fine cause sometimes you want extra things in the content (like the previews), but we lose some DropdownMenu features like arrow navigation. it makes me wonder if we should wrap the MenuGroups inside NavigableToolbar to enable the arrow navigation.

Hey @youknowriad! I thought about arrow navigation and although I didn't implement it in this PR, I had seen the Autocomplete component as some reference [that implements arrow navigation itself ](https://github.com/WordPress/gutenberg/blob/2e814da01dd4c130c8b0cc4fe45732b49736aff5/packages/components/src/autocomplete/index.js#L334).

Could the arrow navigation be implemented with the way you suggest also here?

Maybe others share their thoughts too?

**Comment by ntsekouras** [comment]
> Designwise, I think we should try to move closer to the mockup shown in #18667 (comment):

I thought that was preferred to move Styles below, as mentioned in the issue. Should we revert them as they were before on top?

At first I implemented the lists closer to the first mockup ( more paddings etc ). But as mentioned from @youknowriad and I believe he is right, this wasn't so consistent with the rest vertical lists. After that I made it more compact and to match the hover styles etc.

Also I am wondering if it would make sense to add another level for the user to find out the available Style options, since there is a small number of styles?

Thanks for reviewing @jasmussen! I have also made the css fix you suggested.

**Comment by jasmussen** [comment]
> I thought that was preferred to move Styles below, as mentioned in the issue. Should we revert them as they were before on top?

Ah, thanks for surfacing that:

Screenshot 2020-06-10 at 11 01 58

I think it's fine to try with the styles below for starters. I do think we lost a little bit from this mockup:



A few things can probably get us closer:

- Increase the padding inside both menus to be larger. Let's try `18px` for now — ideally we should use variables that adhere to the 24px grid, but there's some overlap with the inner padding of menu items, and 18px seems like a good start.
- Separator should be `$dark-gray-primary`
- Preview should be spaced better — `top: -$border-width;`, and `left: calc(100% + $grid-unit-15);` seems to work well
- `.block-editor-block-switcher__preview-title` should be `color: $dark-gray-primary`
- `.block-editor-block-switcher__popover .components-popover__content .block-editor-block-switcher__preview` should have 18px padding instead of 10 to match what we're trying above

That gets us this, which I think is a good start:

Screenshot 2020-06-10 at 11 10 53

Let me know if you'd like me to push some changes directly to your branch.

**Comment by ntsekouras** [comment]
> I think it's fine to try with the styles below for starters. I do think we lost a little bit from this mockup:

I had it like this in my first iterations but as pointed out it was inconsistent with other vertical "menus": Do you think that as well designwise?

other vertical

Also @jasmussen please take a look to the changes when smaller viewport. Previously they were just not working well with the Styles Preview and I have placed Preview at the bottom.

> Let me know if you'd like me to push some changes directly to your branch.

Feel free to push them if you have already made them to test the UI. Just let me know. Thank you!

I am now implementing the arrow navigation but since we have two separate MenuGroups and the focus goes to Blocks, arrow navigation will cycle through Blocks only and not the Styles. Any thoughts/workaround on this @youknowriad ?

**Comment by jasmussen** [comment]
Pushed polish that addresses the feedback I left:

Screenshot 2020-06-10 at 12 31 45

There needs to be contrast between menu item section labels (Styles) and the menu items (Default, Large), so I moved from the styling that's present in the global more menu, to the styling used for categories in the inserter. Would be good to make this consistent across all three, with the same gray color.

**Comment by jasmussen** [comment]
I noticed an error in the console:

```
index.js:138 Uncaught TypeError: Cannot read property 'style' of undefined
    at Object.onFrame (index.js:138)
    at update (web.cjs.js:516)
```

**Comment by ellatrix** [comment]
@jasmussen I think that's caused by another PR recently merged with master. I'll have a look.

**Comment by ellatrix** [comment]
When I arrow navigate, the main scroll container scroll up/down.

**Comment by ntsekouras** [comment]
> Would be good to make this consistent across all three, with the same gray color.

@jasmussen do you mean changing the borders here to the same color as Block Switcher?

Screenshot

**Comment by ntsekouras** [comment]
> When I arrow navigate, the main scroll container scroll up/down.

It doesn't happen to me. Can you share your browser info?

**Comment by jasmussen** [comment]
> do you mean changing the borders here to the same color as Block Switcher?

I did not, but we should. Pushed a fix:

Screenshot 2020-06-11 at 10 23 26

Important to only do it for the dark popovers, so we still have the light ones here:

Screenshot 2020-06-11 at 10 23 30

**Comment by ntsekouras** [comment]
Thank you @jasmussen! It's great that you share some details about the design decisions. It's really valuable!! :)

**Comment by jasmussen** [comment]
Always happy to help, ping me anytime. I worked on some of the design implementation of #18667, which isn't actually complete yet, so I'm happy to see every little bit getting us closer to completing it! Thanks again for the PR.

**Comment by ellatrix** [comment]
When I have a block that has both transforms and styles to switch to, I cannot use the arrow keys to jump from transforms to styles. It seems like these two should be the same navigable list?

**Comment by ellatrix** [comment]
Otherwise this looks pretty good when testing!

**Comment by ellatrix** [comment]
Another thing I noticed: the previews of styles may be cut off. It's possible to scroll down, but you have to pass the entire block first because to toolbar sticks to the block. I wonder if the styles may not be better positioned at the top, since it gives more space for previews?

File: packages/block-editor/src/components/block-switcher/index.js
Comment (ellatrix) [inline packages/block-editor/src/components/block-switcher/index.js:226]: This feel like very deep indentation. Is there a possibility to move it elsewhere?

Comment (ellatrix) [reply packages/block-editor/src/components/block-switcher/index.js:226]: (Could be a variable defined earlier, or a separate component, for example.)

File: packages/block-editor/src/components/block-switcher/style.scss
Comment (ellatrix) [inline packages/block-editor/src/components/block-switcher/style.scss:75]: What does this do?

**Comment by ntsekouras** [comment]
> When I have a block that has both transforms and styles to switch to, I cannot use the arrow keys to jump from transforms to styles. It seems like these two should be the same navigable list?

I know and had mentioned it above:

> I am now implementing the arrow navigation but since we have two separate MenuGroups and the focus goes to Blocks, arrow navigation will cycle through Blocks only and not the Styles. Any thoughts/workaround on this @youknowriad ?

I believe that this will require bigger changes to combine these Blocks list and BlockStyles since they are different Components with inner list items and they are both shown conditionally. Feel free to suggest a smaller workaround for that, that I am not aware of.

Would this be a blocker for landing the PR?

**Comment by ellatrix** [comment]
Another small issue: normally popovers adjust the height to fit within the viewport boundary, and the content is scrollable. This doesn't seem to be the case here. This might be caused by the addition of `overflow: visible;`?

Screenshot 2020-06-11 at 14 26 40

File: packages/block-editor/src/components/block-switcher/style.scss
Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/style.scss:75]: This is needed as the Preview changed design implementation, so as to be shown outside the lists Popup.

**Comment by ellatrix** [comment]
> I believe that this will require bigger changes to combine these Blocks list and BlockStyles since they are different Components with inner list items and they are both shown conditionally. Feel free to suggest a smaller workaround for that, that I am not aware of.
> 
> Would this be a blocker for landing the PR?

Thanks for clarifying! I just checked and tab seems to work fine, which means it's possible to navigate to it with the keyboard, so I wouldn't consider it a blocker. :) In fact, it's way worse to navigate the items in `master`...

**Comment by ntsekouras** [comment]
> I wonder if the styles may not be better positioned at the top, since it gives more space for previews?

This was discussed a lot with many people here and at the issue and decided to be at the bottom.

**Comment by ntsekouras** [comment]
> Another small issue: normally popovers adjust the height to fit within the viewport boundary, and the content is scrollable. This doesn't seem to be the case here. This might be caused by the addition of `overflow: visible;`?

Yes it does. I have to check about this!

File: packages/block-editor/src/components/block-types-list/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-types-list/index.js:37]: Will reverse it. Do you want to reverse the `normalizedItems.map` destructure as well?

**Comment by ellatrix** [comment]
> This was discussed a lot with many people here and at the issue and decided to be at the bottom.

Do you have a link to the discussion? Maybe the cutting off wasn't thought about.

**Comment by ntsekouras** [comment]
> Do you have a link to the discussion? Maybe the cutting off wasn't thought about.

here @ellatrix.

issue: https://github.com/WordPress/gutenberg/issues/22073
comment: https://github.com/WordPress/gutenberg/pull/23028#issuecomment-641865366

**Review by ellatrix — APPROVED** [review]
Looks good to me! :)

**Comment by enriquesanchez** [comment]
Hi @ntsekouras, this is a great PR!

I'll scope my feedback to only keyboard interaction, as others have already given you great feedback on visuals and implementation.

I was also expecting to be able to navigate the whole list of options, both Transforms and Styles, with the arrow key. I read your previous point about this not being possible because they are two different kinds of components.

I realize that tabbing works, and that's much better from what we had previously. So definitively not a blocker.

I can't help but wonder if they have to be different components? In terms of visuals and interaction, they now seem to me to be pretty similar. Probably out of scope for this PR, but it'd be a nice follow up to try.

This is important for many reasons, but to give you an example, a blind user who is used to navigating these menus with arrow keys, will simply not know that there is a 'Style' section below 'Transforms'.

**Review by enriquesanchez — APPROVED** [review]

**Comment by ZebulanStanphill** [comment]
Are we okay with how the preview works on mobile? It's worth noting that mobile users usually don't have the ability to hover things, so they won't be able to see the previews.

image

**Comment by jasmussen** [comment]
Before we merge this, let's dwell on the styles treatment for a little bit, this one:



I helped polish that by differentiating the sybheading with the individual styles. But something does get a little bit lost there, compared to what we have today. I think @mtias mentioned at some point that we could create a single “Styles ->” flyout opening up the grid of styles in the popover. To quote him, we want to make sure the design is a "net improvement", and maybe it isn't at the moment.

How much work would it be to revert the styles-as-a-list bit to this again?

Screenshot 2020-06-12 at 08 23 31

Still keeping it at the bottom, and I'd push some polish to make them smaller, perhaps match the efforts that Riad is exploring in https://github.com/WordPress/gutenberg/pull/22789?

Then as a separate PR, we can revisit the styles, to maybe explore that flyout. What do you think?

**Comment by gziolo** [comment]
@youknowriad, you know the API for previews very well, is anything stopping us from adding preview for transforms as well? It would be a super cool addition.

**Comment by ntsekouras** [comment]
Hey @enriquesanchez! Thanks for the feedback!

I agree with you and I am going to make a new issue to handle this combination of both Components ( Blocks and Styles ), as it requires some more work and would be nice to have smaller PRs.

> Are we okay with how the preview works on mobile? It's worth noting that mobile users usually don't have the ability to hover things, so they won't be able to see the previews.

@ZebulanStanphill I agree with you in this one too, but our current preview is not working well, as in smaller viewports shows the right panel preview so small ( maybe like 25px or something at some widths! ) so it's like it's not there..

The main difference in this PR from what we have now, is the small preview Blocks that are currently shown and not anymore.

Screenshot(1)

These Previews are so small that in some Blocks like `Quote`, I do not know if they provide any value and even more maybe "clog" the UI... Of course in other Blocks like `Image` even small the user can get useful information.

So I believe this is an issue mostly about UX and how we want to go in the future after having discussed better all the implications about the smaller viewport Previews ( how to give the most value to end users ), revisit the available Previews for each available Block and probably revisit about which is the best position for Block Styles, top or bottom.

I don't know if we can get our hands on some analytics, if available ( and if not maybe add them ), so as to observe if Block Styles are actually used more frequently than Blocks list. This could help us towards a more informed decision..

File: packages/block-editor/src/components/block-switcher/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-switcher/index.js:167]: Now that this list is different than the list of items displayed in the inserter through `BlockTypesList`, it might be more convenient to avoid all the logic to style it to fit this space by iterating over the list of transforms and using `MenuItem` component for each of them.

Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/index.js:167]: I believe this could be better suited in a new PR with the combining of the two lists ( Blocks and Styles ), as there will be a lot of Component changing/moving around there. But others can provide their input as well..

**Comment by gziolo** [comment]
> We're not using a DropdownMenu here and instead we use just Dropdown, that's fine cause sometimes you want extra things in the content (like the previews), but we lose some DropdownMenu features like arrow navigation. it makes me wonder if we should wrap the MenuGroups inside NavigableToolbar to enable the arrow navigation.

@youknowriad, I guess you meant `NavigableMenu` rather than `NavigableToolbar` here. What's preventing us from using `DropdownMenu` here? It looks like we have two menu groups, both with labels divided with the separator that should be covered solely by using the combination of `MenuGroup` and `MenuItem` components. It shouldn't differ much from other menus that @jasmussen shared in https://github.com/WordPress/gutenberg/pull/23028#issuecomment-642493280.

**Comment by youknowriad** [comment]
> you know the API for previews very well, is anything stopping us from adding preview for transforms as well? It would be a super cool addition.

Nothing, we should just run the transform on a tmp block before showing the preview.

File: packages/block-editor/src/components/block-switcher/style.scss
Comment (mcsf) [reply packages/block-editor/src/components/block-switcher/style.scss:75]: This would no longer be needed once we switched to using Popover, correct?

Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/style.scss:75]: Correct.

File: packages/block-editor/src/components/block-types-list/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-types-list/index.js:63]: Changes in this file are no longer necessary.

File: packages/components/src/dropdown-menu/style.scss
Comment (gziolo) [inline packages/components/src/dropdown-menu/style.scss:86]: Is this class still in use?

Comment (ntsekouras) [reply packages/components/src/dropdown-menu/style.scss:86]: Hey @gziolo . Still in `WIP` status. I will make more changes tomorrow :)

File: packages/block-editor/src/components/block-types-list/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-types-list/index.js:63]: These changes are not used now, but there is just a destructure and added `role` attribute support here. It might be used by the Inserter also to support proper arrow navigation ( there is a check for specific role menuitem ).

Do you think I should revert these changes?

File: packages/components/src/dropdown-menu/style.scss
Comment (jasmussen) [reply packages/components/src/dropdown-menu/style.scss:86]: Yes this class is still in use. `is-alternate` is a prop that should be applied to all popovers inside the editing canvas, so as to provide the same dark borders as the block UI itself.

**Comment by jasmussen** [comment]
I've extracted some of the visual changes into #23192, just to lighten the review burden here. The rebase should be simple.

**Comment by jasmussen** [comment]
Took this for a spin now, this is what I see:

now

Experience wise, the arrow key navigation is really nice now. The vertical list is also solid. 

That alone, is a solid improvement, and if need be, one we could extract and ship on its own.

The weak point is still the styles, as noted in https://github.com/WordPress/gutenberg/pull/23028#issuecomment-643090114. Even though the following is accurate to the initial G2 mockup:

Screenshot 2020-06-16 at 10 23 50

in practice, it does not feel like a net win over what came before:



The preview is a nice addition, but if it's a point of complexity, honestly fine to look at separtely, or even punt entirely. I think what we need to get right is the styles list. 

Riad has a "quick inserter" iteration here: https://github.com/WordPress/gutenberg/pull/22789#issuecomment-641884212

The design for patterns is based on this mockup:



And as [discussed on the thread](https://github.com/WordPress/gutenberg/pull/22789#issuecomment-641884212), the small thumbnails there work well.

Here's what it looks like as compiled from the branch:

Screenshot 2020-06-16 at 10 28 19

If you can revert to the previous style of style previews, I can push to this branch to make them use the same visual treatment as those patterns shown above.

So, potential next steps:

- Postpone the popover preview if it's a point of complexity holding up this PR
- Revert the style change to what was before. Doesn't matter if it's above or below the transforms.
- Polish the style thumbnails to look like the patterns from Riad's new quick inserter (I'll help here)

What do you think?

**Comment by ntsekouras** [comment]
> So, potential next steps:
Postpone the popover preview if it's a point of complexity holding up this PR
Revert the style change to what was before. Doesn't matter if it's above or below the transforms.
Polish the style thumbnails to look like the patterns from Riad's new quick inserter (I'll help here)

Seems okay to me!

**Comment by ellatrix** [comment]
@jasmussen Should the styles be changed back to the 2 column layout? It seems that's closer to the three column layout we have for the blocks.

**Comment by jasmussen** [comment]
> Should the styles be changed back to the 2 column layout? It seems that's closer to the three column layout we have for the blocks.

Yes, I will push some changes when it happens so that it better fits the quick inserter example from Riad's `add/quick-inserter` branch. 

But right now I'm trying to compile that to figure out how the arrow key navigation works there, because as Nik noted that was part of the challenge here.

**Comment by jasmussen** [comment]
Okay, tried @youknowriad's quick inserter branch, and it appears arrow key navigation doesn't work there, it's tab only. 

@gziolo I could really appreciate your advice here. As noted in https://github.com/WordPress/gutenberg/pull/23028#issuecomment-644617837, unfortunately the text-only list for styles isn't working as well as the initial mockup suggested — but the arrow key navigation supposedly is a bit challenged with handling the switch from the vertical list of transforms and going into the horizontal list of styles. Do you have any recommendations on how best to move forward there?

**Comment by jasmussen** [comment]
Took this for a spin. This is what I see:

current state

This is good. It's tabbable, arrow-keyable, and it is a net benefit over what ships in master. There were some issues with the treatment and positioning of the popover for previews, but I have a pending fix for that. I'll push as soon as this is rebased.

tweaks

I also meant to fix so that the style previews looked more like 



however given this PR is big, this is probably best done separately. And it would also be best done in consideration of what's in the sidebar also, so I'll look at that separately.

In the end, I feel like we've arrived at a good place:

- Tabbable, keyboardable
- The vertical list is much better for transforms
- Nice benefit of being able to see previews

It would still be nice to, in the future, follow up on this to explore again the "styles" flyout shown here:



however this is a big undertaking as it needs to consider mobile as well, so this is probably best done separately. And it would be a shame to hold up this improvement for that.

So all this needs now is a code review!

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [inline packages/block-editor/src/components/block-switcher/block-transformations-menu.js:12]: `MenuGroup` takes `label` prop. The way it is implemented now, it won't be announced by screen readers.

https://github.com/WordPress/gutenberg/tree/master/packages/components/src/menu-group#development-guidelines

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [inline packages/block-editor/src/components/block-switcher/block-transformations-menu.js:21]: Did you try wrapping an icon with `BlockIcon` component? 
https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/block-icon

File: packages/block-editor/src/components/block-switcher/index.js
Comment (gziolo) [inline packages/block-editor/src/components/block-switcher/index.js:189]: The same thing with the label here:
```suggestion
											<MenuGroup label={ __( 'Styles' ) }>
```

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:21]: There is similar code in the another file you modified:

```js
icon={ <BlockIcon icon={ icon } showColors /> }
```

**Review by gziolo — COMMENTED** [review]
This is shaping very well, nice improvements. It could be nice to perform some accessibility testing.

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:21]: Yes, but the new parent component used, was expecting a `string` value.

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js]: Thanks! I added both labels.

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:21]: Is anything blocking from changing it to accept component?

Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js]: Changed it to `BlockIcon`. Though there was no problem using it... Probably my initial problem was related to some first iterations

Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js]: @gziolo What exactly is the purpose of the `BlockIcon` component?

File: packages/block-editor/src/components/block-switcher/index.js
Comment (mcsf) [inline packages/block-editor/src/components/block-switcher/index.js:106]: I suggest turning this into a function component and placing it somewhere in this module. That way, we avoid the unnecessary work when we evaluate this expression and don't actually render (`{ hoveredClassName !== null && PreviewBlockPopover }`).

**Review by mcsf — COMMENTED** [review]
This is looking pretty good!

I did experience something that I would guess is related to the switch to Popover. Basically, its flickers when it is taller than the window:

popover-flicker

I vaguely recall similar reports elsewhere, so this might be a tracked issue already.

**Comment by enriquesanchez** [comment]
Hi @ntsekouras 👋

I did another test and the keyboard interaction is so much better. It's great that I can now navigate with both tabs or arrows through the whole popover 🎉

One thing I noticed is that the first time I enter the 'Styles' `group after opening the menu,` I only hear VoiceOver announce the `menuitems`. If I move down to the 'Transforms' `group`, I hear the `aria-labelledby` which lets me know that I'm in the transforms `group`.

If I move back up, I can then hear the 'Styles' `aria-labelledby` when enter that `group` again. 

Any idea why this might be happening?

File: packages/block-editor/src/components/block-styles/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-styles/index.js]: > but for that we might need to revisit how BlockStyles is used inside BlockInspector

I agree with that. For now changed as suggested.

**Comment by ntsekouras** [comment]
Hey @enriquesanchez , thanks for reviewing and your feedback! 

Unfortunately I'm still on my Linux machine and not really sure how to test the voice over, but I'll take a look at it.. From what you're describing, my first guess would be that the Styles menu group is not getting proper focus `onOpen`.

**Comment by youknowriad** [comment]
> I did experience something that I would guess is related to the switch to Popover. Basically, its flickers when it is taller than the window

The flickering is something that was present on master too but now fixed. A rebase might help.

**Comment by ellatrix** [comment]
Great work! Let's merge and iterate if there's some remaining issues. It has some time in master for testing.

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [inline packages/block-editor/src/components/block-switcher/block-transformations-menu.js:25]: Is this class necessary here?

File: packages/dom/src/focusable.js
Comment (gziolo) [inline packages/dom/src/focusable.js:95]: Should the description of the function be updated to reflect changes applied? Does it make sense to include an entry in the CHANGELOG for the package? Is it a bug fix?

Is it possible to cover it with a unit test to prevent further regressions?

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:25]: This class is used in `BlockTypesList` ( it was used before ). I came across some failing e2e tests that were using the `transform` functionality and this class seemed a good idea to preserve, as it might be used by other e2e tests in the future.

Comment (gziolo) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:25]: I don’t think that e2e test should shape production code, and I’m sure you can use accessible labels to achieve the same result as block names are unique. It isn’t a big issue but if we can remove unused code with little effort we should always aim for it.

File: packages/dom/src/focusable.js
Comment (ellatrix) [reply packages/dom/src/focusable.js:95]: I missed this comment. Maybe we can create a follow-up PR?

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:25]: >I don’t think that e2e test should shape production code

I definitely agree with this!

In this specific case though ( `MenuItem` ) there is not any accessible labels to match, besides the content itself. `MenuItem` has an [info property](https://developer.wordpress.org/block-editor/components/menu-item/#info) but this would end up adding stuff again.

So, would it be preferable to search through the content?

File: packages/dom/src/focusable.js
Comment (ntsekouras) [reply packages/dom/src/focusable.js:95]: @gziolo It wasn't exactly a bug, more like making `DropdownMenu`'s arrow navigation work properly by adding the `BlockStyles`, which had nested elements. That was causing problems with the declared `Selectors` here: https://github.com/ntsekouras/gutenberg/blob/2a2b2c8ce042bc06c220eab6891b0d477b6d9740/packages/dom/src/focusable.js#L20 . Nested elements were also selected as `focusables` that caused breaking the navigation.

My first thought was introducing [a new attribute in this commit](https://github.com/WordPress/gutenberg/pull/23028/commits/189b35a7e1751ce0235d4cb56a097ac7b6d3bcee#diff-5f5842dcdf7edec05431ef2f0a79b555R70) ( `focusable.js changes` ), but suggested to simplify it..

I would really appreciate any feedback for how to handle this better..

File: packages/block-editor/src/components/block-switcher/block-transformations-menu.js
Comment (gziolo) [reply packages/block-editor/src/components/block-switcher/block-transformations-menu.js:25]: Well, the content is the accessible label, this is what is going to be announced when users depend on screen readers. By using this approach, we also ensure that UI is accessible. It's harder to target elements by content, but there are many existing tests that do it.

File: packages/dom/src/focusable.js
Comment (gziolo) [reply packages/dom/src/focusable.js:95]: `focusable` is the public API on the `@wordpress/dom` package: https://www.npmjs.com/package/@wordpress/dom. It has 15k downloads per week. It's used by 3rd party projects:
Screen Shot 2020-06-23 at 12 55 49

Using `tabindex="-1"` – it's a well-established pattern used with keyboard navigation:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

> A negative value (usually tabindex="-1") means that the element is not reachable via sequential keyboard navigation, but could be focused with Javascript or visually by clicking with the mouse. It's mostly useful to create accessible widgets with JavaScript.

It feels like it was missed in the initial implementation and it would mean a bug fix. @ellatrix, can you confirm?

Comment (ellatrix) [reply packages/dom/src/focusable.js:95]: It was added by Andrew in #2323. Not sure if iframes were meant to be excluded. I don't really understand the change here. Isn't everything that is tab index -1 already excluded? And an iframe's children can't be queried, so they are also excluded?

Comment (gziolo) [reply packages/dom/src/focusable.js:95]: I think that you need to explicitly mark all children at every nesting level to make it work. It's meant to be addressed with [inert attribute](https://github.com/WICG/inert) but it isn't implemented in all browsers yet.

**Pull request merged**
